File: src/main/java/thedarkcolour/futuremc/item/ItemDye.java
package thedarkcolour.futuremc.item;

import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.passive.EntitySheep;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.NonNullList;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import thedarkcolour.core.item.Modeled;
import thedarkcolour.futuremc.FutureMC;
import thedarkcolour.futuremc.init.FutureConfig;

public class ItemDye extends Item implements Modeled {
    public ItemDye() {
        setHasSubtypes(true);
        setMaxDamage(0);
        setTranslationKey(FutureMC.ID + "." + "dye");
        setRegistryName("dye");
        setCreativeTab(FutureConfig.general.useVanillaTabs ? CreativeTabs.MISC : FutureMC.TAB);
        addModel();
    }

    @SideOnly(Side.CLIENT)
    public void model() {
        ModelLoader.setCustomModelResourceLocation(this, 0, new ModelResourceLocation(getRegistryName() + "white", "inventory"));
        ModelLoader.setCustomModelResourceLocation(this, 1, new ModelResourceLocation(getRegistryName() + "blue", "inventory"));
        ModelLoader.setCustomModelResourceLocation(this, 2, new ModelResourceLocation(getRegistryName() + "brown", "inventory"));
        ModelLoader.setCustomModelResourceLocation(this, 3, new ModelResourceLocation(getRegistryName() + "black", "inventory"));
    }

    @Override
    public String getTranslationKey(ItemStack stack) {
        return getTranslationKey() + "." + stack.getMetadata();
    }

    @Override
    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
        if(this.isInCreativeTab(tab)) {
            for(int i = 0; i < 4; ++i) {
                items.add(new ItemStack(this, 1, i));
            }
        }
    }

    @Override
    public boolean itemInteractionForEntity(ItemStack stack, EntityPlayer playerIn, EntityLivingBase target, EnumHand hand) {
        if(target instanceof EntitySheep) {
            EntitySheep entityIn = (EntitySheep) target;
            EnumDyeColor color = getColor(stack);

            if(entityIn.getFleeceColor() != color) {
                entityIn.setFleeceColor(color);
                stack.shrink(1);
                return true;
            }
        }
        return false;
    }

    public EnumDyeColor getColor(ItemStack stack) {
        if(stack.getMetadata() > 3) return EnumDyeColor.WHITE;
        switch (stack.getMetadata()) {
            case 1: return EnumDyeColor.BLUE;
            case 2: return EnumDyeColor.BROWN;
            case 3: return EnumDyeColor.BLACK;
            default: return EnumDyeColor.WHITE;
        }
    }
}

File: src/main/java/thedarkcolour/futuremc/item/ItemDye.kt
package thedarkcolour.futuremc.item

import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.passive.EntitySheep
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumHand
import net.minecraft.util.NonNullList
import thedarkcolour.core.util.setItemModel
import thedarkcolour.core.util.setItemName
import thedarkcolour.futuremc.FutureMC
import thedarkcolour.futuremc.config.FConfig

class ItemDye : Item() {
    init {
        setItemName(this, "dye")
        setHasSubtypes(true)
        maxDamage = 0
        creativeTab = if (FConfig.useVanillaCreativeTabs) CreativeTabs.MISC else FutureMC.TAB

        addModels()
    }

    private fun addModels() {
        setItemModel(this, 0, registryName.toString() + "white")
        setItemModel(this, 1, registryName.toString() + "blue")
        setItemModel(this, 2, registryName.toString() + "brown")
        setItemModel(this, 3, registryName.toString() + "black")
    }

    override fun getTranslationKey(stack: ItemStack): String {
        return translationKey + "." + stack.metadata
    }

    override fun getSubItems(tab: CreativeTabs, items: NonNullList<ItemStack>) {
        if (isInCreativeTab(tab)) {
            for (i in 0..3) {
                items.add(ItemStack(this, 1, i))
            }
        }
    }

    override fun itemInteractionForEntity(
        stack: ItemStack,
        playerIn: EntityPlayer,
        target: EntityLivingBase,
        hand: EnumHand
    ): Boolean {
        if (target is EntitySheep) {
            val color = getColor(stack)
            if (target.fleeceColor != color) {
                target.fleeceColor = color
                stack.shrink(1)
                return true
            }
        }
        return false
    }

    private fun getColor(stack: ItemStack): EnumDyeColor {
        return if (stack.metadata > 3) EnumDyeColor.WHITE else when (stack.metadata) {
            1 -> EnumDyeColor.BLUE
            2 -> EnumDyeColor.BROWN
            3 -> EnumDyeColor.BLACK
            else -> EnumDyeColor.WHITE
        }
    }
}